api_compatibility_policy.md

```diff
@@ -43,21 +40,83 @@ alpha:
 * `Pipeline`
 * `PipelineRun`
 
-We are following [the Kubernetes definitions of these stages](https://kubernetes.io/docs/reference/using-api/api-overview/#api-versioning)
-and we are following [the Kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
-
-* Alpha:
-  * These features may be dropped at any time, though you will be given at least
-    one release worth of warning.
-* Beta:
-  * These features will not be dropped, though the details may change.
-  * Any [backwards incompatible changes](#backwards-incompatible-changes) must be
-    introduced in a backwards compatible manner first, with a deprecation warning
-    in the release notes and migration instructions.
-  * You will be given at least 9 months to migrate before a backward incompatible
-    change is made. This means an older beta API version will continue to be
-    supported in new releases for a period of at least 9 months from the time a
-    newer version is made available.
+### Alpha CRDs
+
+- For Alpha CRDs, the `apiVersion` contains the `alpha` (e.g. `v1alpha1`)
+  
+- Features may be dropped at any time, though you will be given at least one release worth of warning.
+
+### Beta CRDs
+
+- The `apiVersion` field of the CRD contains contain `beta` (for example, `v1beta1`).
+
+- These features will not be dropped, though the details may change.
+
+- Any [backwards incompatible changes](#backwards-incompatible-changes) must be introduced in a backwards compatible manner first, with a   deprecation warning in the release notes and migration instructions.
+
+- Users will be given at least 9 months to migrate before a backward incompatible change is made. This means an older beta API version will continue to be supported in new releases for a period of at least 9 months from the time a newer version is made available.
+
+- Beta API versions follow the [Go libraries compatibility policy](#go-libraries-compatibility-policy).
+
+### GA CRDs
+
+- The `apiVersion` field of the CRD is `vX` where `X` is an integer.
+
+- Stable API versions remain available for all future releases within a Tekton major version.
+
+- Stable features may be marked as deprecated but may only be removed by incrementing the api version (i.e v1 to v2).
+
+- We will not make any backwards incompatible changes to fields that are stable without incrementing the api version.
+
+- Stable APIs do not follow the [Go libraries compatibility policy](#go-libraries-compatibility-policy) i.e. backwards incompatible changes to the Go struct is not allowed.
```

Comment:
Hmm, I was thinking of only a subset of the go packages - esp. the ones in pkg/api/v1 and pkg/client. I'm not sure we need to have all exported functions under this

Reply:
Honestly maybe everything that's not in apis should be in an internal package?

Reply:
> Honestly maybe everything that's not in apis should be in an internal package?

In theory yes 😬. Like in theory most of what is in `pkg` today should be in `internal/pkg` or something unless required otherwise (like `pkg/apis`, the generated clients, …). In practice using `internal` is not always super known by go dev.

Reply:
@dibyom I'd say anything visible in https://pkg.go.dev/github.com/tektoncd/pipeline is fair game.

+1 for moving things to internal.

Reply:
Any reason not to move things to internal? Vincent it sounds like you're saying we haven't done this just because most people aren't aware of the option? If that's the case I'm happy to make this change

Reply:
tracked in https://github.com/tektoncd/pipeline/issues/5679.

Reply:
There is more to it than just "move" though. We did not use it too much initially, because it's not that common to use in the OSS golang community. But now, the project is there for a couple of years already, there is probably quite some project depending on some parts of our go packages, even only in the `tektoncd` org (triggers, …). We need to make sure we are not doing a "switch off" on those and, for each package, look at what is exported, why, is it useful, and is it used elsewhere ?

Reply:
At least for triggers, we only use pkg/client and pkg/apis (we have one test relying on pkg/names but that can probably be fixed easily). I do agree with your statement @vdemeester. My initial suggestion was just keeping only pkg/apis and pkg/client under the API policy - I don't think its feasible for us to promise no breaking changes to all of pkg/

Reply:
I agree with Billy that things indexed on pkg.go.dev make sense to cover in our compatibility policy. I would find it confusing to have a compatibility policy that covers some of our public go packages but not others. I agree we don't want to promise no changes to pkg/, but we could say we won't remove/rename a deprecated function/const for x time?

Reply:
> we could say we won't remove/rename a deprecated function/const for x time?

We could perhaps provide that but backwards compatibility technically also includes things like adding new arguments or return values to a function: https://go.dev/blog/module-compatibility 

@wlynch I'm not sure how we can both guarantee go module compatibility while also making rapid changes to alpha APIs if all of pkg/ is under scope.

I think clearly stating out which packages are in scope and which are not is a decent middle ground. Over time, we could move the the out of scope packages into internal. Another option would be to clearly write out what subset of compatibility we can provide e.g. we can support not deprecating/removing functions but we'd probably want the ability modify function signatures in most of pkg/

Reply:
I'm a bit confused about how the policy can be different for stable APIs only - as go libraries are not necessarily bound to a specific API version. 

Does this summary make sense:
- go APIs specific to stable Tekton APIs are subject to the new policy, i.e. we cannot do backwards incompatible changes. Also, a future `v2` public API must be separated in code from the `v1` public API
- go APIs specific to alpha/beta Tekton APIs can be subject to backwards-incompatible changes
- go APIs which are shared between alpha/beta and stable that needs to stay public would have to be separated,  one copy under the stable tree, one copy under the alpha/beta one
- go APIs that we don't need to be public would have to become private, and we could (1) move them to private (2) create a public shim (3) deprecate the shim right away. This would have to happen before v1

Reply:
The first two points make sense to me. 

- For public go apis shared between alpha and beta - why do we need to copy unless we are making a breaking change? There is quite a bit of complexity involved in this so I'm sure we should do this unless we really need to

- I support moving them to private incrementally but I don't think we need to block the v1 API release on this. I'm unsure why we'd need to deprecate the shim.

We had some discussion on this in the last API WG - the conclusion was that I'd update the policy to mention add  guidelines for making changes to the go libraries. Specifically this would be to avoid making backwards incompatible changes to the go APIs pertaining to the V1 API version.

Reply:
> The first two points make sense to me.
> 
> * For public go apis shared between alpha and beta - why do we need to copy unless we are making a breaking change? There is quite a bit of complexity involved in this so I'm sure we should do this unless we really need to

I meant shared between alpha/beta and stable. If we can make sure (through testing) that we would catch those cases, then it would be fine to not copy the code until a breaking change is needed on alpha or beta side.

> * I support moving them to private incrementally but I don't think we need to block the v1 API release on this. I'm unsure why we'd need to deprecate the shim.

We don't need to block v1 only on this as long as our API stability policy excludes them upfront. 
If we create a shim, it would only be to give existing users more time to move away from those APIs that we want to become private. But we would have to deprecate the shim to signal our users that that is only a temporary solution.
Alternatively we could deprecate the API before making it private, and when it moves to private provide no shim at all.

> 
> We had some discussion on this in the last API WG - the conclusion was that I'd update the policy to mention add guidelines for making changes to the go libraries. Specifically this would be to avoid making backwards incompatible changes to the go APIs pertaining to the V1 API version.